src/vm/contexts.rs

```diff
@@ -742,23 +765,45 @@ impl<'a, 'b> Environment<'a, 'b> {
                 panic!("Tried to construct environment with bad caller {}", caller);
             }
         }
+        if let Some(ref sponsor) = sponsor {
+            if let Value::Optional(_) = sponsor {
+                let inner = Value::expect_optional(sponsor.clone());
```

Comment:
Minor nitpick (not a bug) -- the `Value::expect_*` method will all print out panic messages if their arguments don't have the right types.  Up to you if you want to use them here instead.

Reply:
Good to know. I think I may leave it as is here because I pattern-matched right before calling.